Model noise

SRFU-NN opened the pull request:
Separate system for modelling noise

SRFU-NN committed:
feat: Model system noise in separate class

SRFU-NN committed:
doc: alligningn doc style with standard

SRFU-NN committed:
feat: New default settings for MultiplicativeNoise

SRFU-NN committed:
refactor: Collected evaluation of random noise

SRFU-NN committed:
feat: Reproducible "random" tests

SRFU-NN committed:
test: Tests for new default settings

SRFU-NN committed:
test: Testing examples in docstring

SRFU-NN committed:
style: removed superfluous whitespace

SRFU-NN committed:
fix: Python 3.8 does not like list[float]

typing.List[float] is used instead of list[float]

SRFU-NN committed:
feat: removed nuderlying_noise_model

SRFU-NN committed:
feat: Switched from "apply" to "get_noise"

SRFU-NN committed:
feat: Removed NoiseModel.apply()

SRFU-NN committed:
doc: updated docstring

Now no longer references noise_dist, which shouldn't be directly
accessed.

SRFU-NN committed:
refactor: Renamed "noise" to "raw_noise"

SRFU-NN committed:
test: Testing inaccessability of raw_noise

SRFU-NN committed:
feat: Added SumNoise class

SRFU-NN committed:
fix: DataDependentNoise.raw_noise inaccessible

SRFU-NN committed:
feat: set_noise_type

SRFU-NN committed:
feat: Noise model factory

SRFU-NN committed:
test: parse_noise_model

SRFU-NN committed:
test: Imports

SRFU-NN committed:
style: Line length

SRFU-NN committed:
fix: Imports

SRFU-NN committed:
fix: Missed one renaming

SRFU-NN committed:
doc: Justification for noise_size

SRFU-NN committed:
style: Multiline docstring starting on a new line

SRFU-NN committed:
fix: Updated ModelSystem.set:noise_model()

SRFU-NN commented on ProcessOptimizer/model_systems/noise_models.py:
```diff
@@ -0,0 +1,157 @@
+                            f"{self.__class__.__name__} is not supposed to be called.")
+        return self._noise_distribution()*self.noise_size
+    
+    def set_noise_type(self,noise_type: str):
```
Hvilke støj-former har vi brug for? hvis vi ikke har brug for uniform kan det gøres lidt lettere, men måske det er værd at gemme koden for nemt at kunne lave flere i fremtiden?

SRFU-NN commented on ProcessOptimizer/model_systems/noise_models.py:
````diff
@@ -0,0 +1,157 @@
+    noise_model = DataDependentNoise(noise_models=noise_choice)
+    ```
+    """
+    def __init__(self, noise_models: Callable[...,NoiseModel], **kwargs):
````
noise_models er ikke det bedste navn. har I nogle bedre bud (generelt er variabel-navne i flertal et rødt flag)

dk-teknologisk-mon replied:
Some ideas: noise_equation, noise_mapping, noise_function?

SRFU-NN replied:
noise_function er købt, der er jo en funktion.

Og så skal vi have lavet en klasse der kører på mappings, det har den fordel at de kan laves på forhånd, og at den funktion der vælger bare skal spytte en string ud for hvert punkt, så den bliver simplere at definere. Men det blive ikke i denne omgang.

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/model_system.py:
```diff
@@ -56,5 +51,34 @@ def result_loss(self, result):
+    def set_noise_model(self, noise_model: Union[str,dict,NoiseModel], **kwargs):
+        """Sets the noise model for the model system
+
+        Args:
```
In the rest of the codebase we use the word "Parameters", so for consistency we should probably stick to this here also.

SRFU-NN replied:
Dammit, I thought I had changed all of them!

SRFU-NN committed:
doc: typo

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/noise_models.py:
```diff
@@ -0,0 +1,157 @@
+        return self.raw_noise
+
+    
+class MultiplicativeNoise(NoiseModel): # Should this be named ProportionalNoise?
```
I like ProportionalNoise better as it fits with the description that follows.

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/noise_models.py:
```diff
@@ -0,0 +1,157 @@
+            raise ValueError(f"Noise distribution \"{noise_type}\" not recognised.")
+
+
+class AdditiveNoise(NoiseModel): # Should this be named ConstantNoise?
```
I like ConstantNoise better.

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/noise_models.py:
```diff
@@ -0,0 +1,157 @@
+
+class AdditiveNoise(NoiseModel): # Should this be named ConstantNoise?
+    """
+    Noise model for constant noise.
```
Could make it slightly clearer by rewording to: "Noise model for noise that is constant across the entire system. This is most typical for measurement noise."

SRFU-NN replied:
Good idea to have typical use cases.

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/noise_models.py:
```diff
@@ -0,0 +1,157 @@
+    Parameters:
+    * `noise_size` [float, default 0.01]: The size of the noise realitve to the signal.
+    """
+    def __init__(self, noise_size : float = 0.01,**kwargs):
```
The default value represents a dream scenario from the point of view of most companies I've worked with. I'd prefer it if the default (that people may be temped to use in their benchmarks) was slightly larger, either 0.05 or 0.1.

SRFU-NN replied:
I will set it to 0.1, as I prefer erring on the side of grimness with defaults. It is more likely people will discover that their noise is too large than that they will discover that it is too small.

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/noise_models.py:
```diff
@@ -0,0 +1,157 @@
+        pass
+    
+    @property
+    def raw_noise(self) -> float:
```
Could a more intuitive function name be "sample_noise"?

SRFU-NN replied:
Changed, and appended an underscore to mark that it should not be accessed from outside the instance.

SRFU-NN committed:
fix: Restting noise model list

When defining a new noise model list, the old should be discarded.

SRFU-NN committed:
doc: Updaed comment on noise_size

SRFU-NN commented on ProcessOptimizer/model_systems/noise_models.py:
```diff
@@ -0,0 +1,158 @@
+        if noise_type in ["normal", "Gaussian", "norm"]:
+            self._noise_distribution = norm.rvs
+        elif noise_type == "uniform":
+            self._noise_distribution = uniform(loc=-1,scale=2).rvs
```
Is uniform noise of "size" on from -1 to 1 (size is the maximum value of the noise), or from -0.5 to 0.5 (size is the maximum difference between two noise measurements)?

dk-teknologisk-mon replied:
This is pure gut feeling, but this implementation of values from -1 to +1 reflects what I would expect as defaults for a constant noise scheme centered on zero. At least more so than -0.5 to 0.5. I expect this noise type will see relatively little use compared to normally distributed noise, so I don't think the choice matters that much.

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/noise_models.py:
```diff
@@ -0,0 +1,159 @@
+class SumNoise(NoiseModel):
+    """
+    Noise model that returns the sum of two or more noise models. Can be used if the
+    total noise is a sum of constant and relative noise.
```
... "is a sum of constant and propotional noise." to use same wording as the other methods.

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/model_system.py:
```diff
@@ -8,29 +11,20 @@ class ModelSystem:
-                # The score of the system is returned
-                return score
+        Function for calculating the noiseless score of the system at a given point in 
+        the parameter space. It is expected to have the following signature.
```
I'm not against removing the long description of the structure of the score, but if we do so, we should also kill the last sentence (It is expected to have the following signature.).

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/model_system.py:
```diff
@@ -8,29 +11,20 @@ class ModelSystem:
+
     """
-    def __init__(self, score, space, true_min=None):
+    def __init__(self, score: Callable[..., float], space, true_min=None, noise_model: NoiseModel = ZeroNoise):
```
It is not super important, but I would prefer if the default assumption is that noise exists and has to be considered (aka. a ConstantNoise with default values at least). I should hope that our benchmarks use this case more often than the ZeroNoise assumption :)

SRFU-NN replied:
I don't like having ConstantNoise as a default, because the default noise size will likely be too small or too large. 

Perhaps we should just remove the default?

Anyway, I will pipe it through set_noise_model(), so we get more options to define the noise model.

dk-teknologisk-mon replied:
I think it makes sense to remove the default and force users to explicitly decide what they want to do. They have to do so for the space-parameter anyway.

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/model_system.py:
```diff
@@ -8,29 +11,20 @@ class ModelSystem:
-        The true minimum value of the score function within the parameter space. 
+        The true minimum value of the score function within the parameter space.
+
+    * `noise_model` [NoiseModel]:
```
This parameter has a default, which we should state here.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_noise_model.py:
```diff
@@ -0,0 +1,200 @@
+def test_additive_noise(signal_list):
+    noise_model = AdditiveNoise()
+    noise_model._noise_distribution = lambda: 2
+    noise_list = [noise_model.get_noise(None,signal)for signal in signal_list]
```
Missing a space inside the list comprehension, just before the "for".

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_noise_model.py:
```diff
@@ -0,0 +1,200 @@
+    noise_list = [noise_model.get_noise(input,signal) for signal in signal_list]  
+    assert [noise==signal for (signal,noise) in zip(signal_list,noise_list)]
+
+def test_zero_noice(signal_list):
```
Typo in test name, should read "test_zero_noise"

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_noise_model.py:
```diff
@@ -0,0 +1,200 @@
+        noise_model.raw_noise
+
+# raw_noise of AdditiveNoise should not be accesible, since it is a compound NoiseModel,
+# and should ot have "its own" noise.
```
Is the test here (test_data_dependent_raw_noise_error) a left-over from building the test of the SumNoise class? Because with the way you have built AdditiveNoise (ConstantNoise) now, it should definitely have its own noise? Or am I misunderstanding what this test should do?

SRFU-NN replied:
AdditiveNoise have its own noise. SumNoise does not. It would probably be clearer is SumNoise summed over two noise sources, and not just one, that is an actual use case, where this is just "however little I can make it work with".

SRFU-NN replied:
Wait, I see the issue, I wrote the wrong noise type in the comment. That must have been confusing. I will change the comment

dk-teknologisk-mon reviewed (commented):
Very nice additions to the structure of the ModelSystem class. I have added minor comments/suggestions throughout.

SRFU-NN committed:
feat: Renamed noise models

additive noise is now called constant noise, multiplicative noise is now
called proportional noise.

SRFU-NN committed:
feat: New default noise size for ProportionalNoise

SRFU-NN committed:
doc: Editorial changes

SRFU-NN committed:
feat: More options for the noise in ModelSystem

Piping the input noise_modle through set_noise_model makes the __init__
function moe flexible.

SRFU-NN committed:
style: typos and other editiorial changes

SRFU-NN committed:
style: renamed raw_noise to _sample_noise

SRFU-NN committed:
feat: Removed default noise model from ModelSystem

Most often, we would like constant noise, but we don't know the size. it
is safer to not set any by default.

SRFU-NN committed:
style: Renamed noise_models() to noise_function()

SRFU-NN committed:
style: Renamed noise_models() to noise_function()

SRFU-NN committed:
fix: Error in setting noise_model in ModelSystem.__init__()

SRFU-NN committed:
fix: A bit too aggresive when changing variable name

sqbl commented on ProcessOptimizer/model_systems/model_system.py:
```diff
@@ -8,29 +11,29 @@ class ModelSystem:
+        the parameter space.
 
     * `space` [Space]:
         The parameter space in the form of a Space object. 
```
"The parameter space in the form of a Space object". 
Because I'm relatively known in the package, I understand this as being an instantiated object of a space. As in `opt.space`. 
I believe that the intent is for the user to give a list of lists (or tubles) as is done during the initiation of the "opt"-object.

SRFU-NN replied:
Good point, I am going to make a space_factory() function in the space module, so we have the option of not caring whether we give the model system a space of the definitions of a space. This will also be useful in many other places, so it is worth having with space, and not in model_system.

sqbl commented on ProcessOptimizer/model_systems/model_system.py:
```diff
@@ -39,6 +42,7 @@ def __init__(self, score, space, true_min=None):
             scores = [score(point) for point in points]
             true_min = np.min(scores)
         self.true_min = true_min
+        self.set_noise_model(noise_model)
```
I'm not really very strong in initiation of classes. 
Would it not be simpler to skip the "set_noise_model" method of the class and then just do this during the initialization:
`self.noise_model = parse_noise_model(noise_model, **kwargs)`?

SRFU-NN replied:
Then we couldn't change the noise model later, which me might want to. But it is a valid question whether we would ever do that.

SRFU-NN replied:
It is there so you can easily change the noise model after initialisation, but that is probably more functionality than we want.  I will remove it.

sqbl commented on ProcessOptimizer/model_systems/noise_models.py:
```diff
@@ -0,0 +1,163 @@
+    the score. This is typical of some electronic measurements.
+
+    Parameters:
+    * `noise_size` [float, default 0.01]: The size of the noise relative to the signal.
```
default in doc-string says 0.01, code says 0.1

SRFU-NN replied:
I forgot to change the docstring when updating the default value, changed.

SRFU-NN committed:
Merge remote-tracking branch 'origin/model_noise' into model_noise

SRFU-NN committed:
feat: space_factory

SRFU-NN committed:
doc: Updated docstring for ProportionalNoise

SRFU-NN committed:
feat: Removed ModelSystem.set_noise_model()

sqbl commented:
LGTM, I'll let Morten have a final say

dk-teknologisk-mon reviewed (approved):
Everything looks good to me! Really nice addition to the functionality that should make it a lot easier to build benchmarks in the near future.
